### ISSUE TITLE
Refactor WorkflowManager

### ISSUE THREAD

**Issue by dpfaffenbauer** [pimcore/pimcore#1653]
Workflow is a great feature of Pimcore, but working with it from PHP API is a pain in the ass. The current WorkflowManager is designed to only really work with the Backend UI when a User is actively logged into Pimcore Admin, as well as it could need some refactoring to make it easier understandable and more adaptable.

For example: Whenever I do a State change within PHP, i need to do following to make it work:

```php
        $user = \Pimcore\Model\User::getById(0);
        $proxyUser = new User($user);
        $manager = \Pimcore\WorkflowManagement\Workflow\Manager\Factory::getManager($proposal, $user);

        /**
         * This is so stupid, Pimcores Workflow needs a valid token storage :/.
         */
        $originalUser = $this->tokenStorage->getToken()->getUser();
        $this->tokenStorage->getToken()->setUser($proxyUser);

        $params['action'] = $action;

        if ($manager->validateAction($params['action'], $params['newState'], $params['newStatus'])) {
            try {
                $manager->performAction($params['action'], $params);
                \Pimcore\Logger::debug('CoreShop State update. ID: '.$proposal->getId().', newState: "'.$params['newState'].'", newStatus: "'.$params['newStatus'].'"');
            } catch (\Exception $e) {
                throw new \Exception('changeOrderState Error: '.$e->getMessage());
            }
        } else {
            throw new \Exception('changeOrderState Error: '.$manager->getError());
        }

        $this->tokenStorage->getToken()->setUser($originalUser);
```

So, what this does is following:
 - Get Pimcore System User (ID 0)
 - Proxy the Pimcore-User, into the Pimcore-Symfony-Security User object
 - Backup my existing Frontend User
 - Set Pimcore-Symfony-Security into the Token Storage
 - Perform my action
 - Backup Original Frontend User into the Token Storage

The reason why this needs to be done this is way is: https://github.com/pimcore/pimcore/blob/master/pimcore/lib/Pimcore/WorkflowManagement/Workflow/Manager.php#L606

```php
$this->element->isAllowed('publish'));
```

actually does

```
 $currentUser = \Pimcore\Tool\Admin::getCurrentUser();
```

which gets the TokenStorageUserResolver (which by the way has a Container dependency, which should be refactored as well, I'll provide a PR for that) which resolves the User from the Token Storage.

So, changing a Workflow State from Code is a huge pain in the ass, it creates stupid and hard to read code.

I mean, I could create a PR with a refactor, but I think it's quite a huge change and should be done by the Core Team.

PS: Since Pimcore is bringing a new major, I guess you could git rid of this file as well:
https://github.com/pimcore/pimcore/blob/master/pimcore/lib/Pimcore/WorkflowManagement/Workflow.php

**Comment by brusch** [comment]
@cleggypdc What's your opinion on that?

**Comment by dpfaffenbauer** [comment]
@brusch @cleggypdc Maybe we can take advantage of already existing stuff for Symfony
 - https://github.com/winzou/StateMachineBundle
 - https://symfony.com/doc/current/components/workflow.html

Only thing is, these two options don't have a separation for States and Statuses. Which is in terms of BC a real problem :/ But: maybe there is a solution for that as well.

**Comment by cleggypdc** [comment]
@brusch yes totally agree its a real pain to manage programatically, the workflow manager was implemented for front end only and needs refactoring for achieving things through the API - it's been on our roadmap for a while so that we can acheive things like programatic state changes and bulk updates.

On the other hand if you are doing something to do with workflow programatically anyway, why not just use the API to change the workflow state and then add a note or fire an event.

**Comment by dpfaffenbauer** [comment]
@cleggypdc Because we don't want to have unnecessary duplicated business logic and code in our application. If pimcore decides to change something regarding workflows, I always would need to adapt my programmatic approach as well. Therefore its kinda nice to have a WorkflowManager which can be called programmatically as well ;)

**Comment by dpfaffenbauer** [comment]
There is no real intention here of changing anything from the Pimcore team?

My Idea would be to take advantage of Symfony Workflow Component. But: one thing it that symfony workflow doesn't support state/status combination. it only allows one state.

Now the question: Deos somebody really use differentiation of states and statuses. IMO: nobody really understands what the difference really is.

The advantage of the workflow component should be clear, it integrates nicely into symfony stack and doesn't add any overhead. It already supports different formats of configuration and pimcore basically just needs to add a GUI to handle the workflow. (As well as some permission checking, right?)

**Comment by alexpozzi** [comment]
Is there any update on this issue? 
I'm making some tests for a prospect where the workflow is partially managed by an external system and I agree on every comment made by @dpfaffenbauer.
It would be nice as well to have the possibility to define multiple workflows for the same object class and decide which workflow start depending on the user needs.

**Comment by dpfaffenbauer** [comment]
@alexpozzi We decided to not use pimcore workflow anymore as its unpracticable in my use cases. We  now use symfony workflow and it works quite well. We also created a separated bundle cause symfony-workflow comes with some limitations: (like you can't extend workflows using configuration) https://github.com/coreshop/WorkflowBundle

**Comment by alexpozzi** [comment]
Thanks @dpfaffenbauer , I'll have a look at it!

**Comment by brusch** [comment]
Closed in favor of #3166
